File: scrapers/naic.py
# NAIC SOLAR — Fallback scraper for states without dedicated scrapers
# Source: https://sbs.naic.org/solar-external-lookup/
#
# NAIC SOLAR is a national database that covers ALL states.
# Slower and less detailed than state-specific lookups, but works everywhere.
# Supports: first name + last name + optional state filter.

import logging
from typing import Optional

import aiohttp
from bs4 import BeautifulSoup

from scrapers.base import LicenseResult, StateScraper

logger = logging.getLogger(__name__)

SOLAR_URL = "https://sbs.naic.org/solar-external-lookup/"


class NAICScraper(StateScraper):
    """Fallback scraper using NAIC SOLAR for any state."""

    STATE_CODE = "ALL"
    STATE_NAME = "NAIC (National)"
    LOOKUP_URL = SOLAR_URL

    def __init__(self, state_code: str = ""):
        self._state_code = state_code
        self._session: Optional[aiohttp.ClientSession] = None

    async def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None or self._session.closed:
            self._session = aiohttp.ClientSession(
                headers={
                    "User-Agent": (
                        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                        "AppleWebKit/537.36 (KHTML, like Gecko) "
                        "Chrome/120.0.0.0 Safari/537.36"
                    ),
                }
            )
        return self._session

    async def lookup_by_name(
        self, first_name: str, last_name: str
    ) -> list[LicenseResult]:
        """Search NAIC SOLAR by name."""
        try:
            session = await self._get_session()

            # First GET to establish session/cookies
            async with session.get(SOLAR_URL) as resp:
                if resp.status != 200:
                    return [LicenseResult(error=f"HTTP {resp.status}")]

            payload = {
                "firstName": first_name.strip(),
                "lastName": last_name.strip(),
            }
            if self._state_code:
                payload["state"] = self._state_code.upper()

            async with session.post(SOLAR_URL, data=payload) as resp:
                if resp.status != 200:
                    return [LicenseResult(error=f"HTTP {resp.status}")]
                html = await resp.text()

            return self._parse_results(html)

        except Exception as e:
            logger.error(f"NAIC SOLAR lookup error: {e}")
            return [LicenseResult(error=str(e))]

    async def lookup_by_npn(self, npn: str) -> list[LicenseResult]:
        """NAIC SOLAR doesn't support direct NPN search via the web form."""
        logger.warning("NAIC SOLAR web form doesn't support NPN search")
        return [LicenseResult(found=False, error="Use name search for NAIC")]

    async def lookup_by_license_number(
        self, license_number: str
    ) -> list[LicenseResult]:
        """NAIC SOLAR doesn't support license number search."""
        return [LicenseResult(found=False, error="Use name search for NAIC")]

    def _parse_results(self, html: str) -> list[LicenseResult]:
        """Parse NAIC SOLAR results."""
        soup = BeautifulSoup(html, "html.parser")
        results = []

        table = soup.find("table", class_="results") or soup.find("table")
        if not table:
            if "no results" in html.lower() or "not found" in html.lower():
                return [LicenseResult(found=False)]
            return [LicenseResult(found=False)]

        rows = table.find_all("tr")
        if len(rows) < 2:
            return [LicenseResult(found=False)]

        for row in rows[1:]:  # Skip header
            cells = row.find_all("td")
            if len(cells) < 3:
                continue

            result = LicenseResult(
                found=True,
                state=self._state_code or "",
            )

            try:
                result.npn = cells[0].get_text(strip=True)
                result.full_name = cells[1].get_text(strip=True)
                if len(cells) > 2:
                    result.state = cells[2].get_text(strip=True)
                if len(cells) > 3:
                    result.license_number = cells[3].get_text(strip=True)

                # NAIC doesn't always show status in search results
                # Mark as active if found (needs detail check for confirmation)
                result.active = True
                result.status = "Found in NAIC"

            except (IndexError, AttributeError) as e:
                logger.warning(f"NAIC parse error: {e}")
                continue

            results.append(result)

        return results if results else [LicenseResult(found=False)]

    async def close(self) -> None:
        if self._session and not self._session.closed:
            await self._session.close()

File: scrapers/naic.go
package scrapers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
	tls_client "github.com/bogdanfinn/tls-client"
)

// SessionFactory creates a new TLS client session with isolated cookies.
type SessionFactory func() (tls_client.HttpClient, error)

// naicAPIResponse matches the JSON structure from the NAIC SBS API.
type naicAPIResponse struct {
	Name                  string      `json:"name"`
	NPN                   string      `json:"npn"`
	LicenseNumber         json.Number `json:"licenseNumber"`
	LicenseType           string      `json:"licenseType"`
	LicenseTypeCode       string      `json:"licenseTypeCode"`
	LicenseEffectiveDate  string      `json:"licenseEffectiveDate"`
	LicenseExpirationDate string      `json:"licenseExpirationDate"`
	LOAs                  string      `json:"loas"`
	Residency             string      `json:"residency"`
	BusinessAddress       string      `json:"businessAddress"`
	BusinessPhone         string      `json:"businessPhone"`
}

// naicErrorResponse matches the NAIC error JSON.
type naicErrorResponse struct {
	UnexpectedError string `json:"UnexpectedError"`
}

// NAICScraper queries the NAIC SBS API for license information.
type NAICScraper struct {
	stateCode      string
	sessionFactory SessionFactory
}

// NewNAICScraper creates a new NAIC scraper for the given state code.
func NewNAICScraper(sessionFactory SessionFactory, stateCode string) *NAICScraper {
	return &NAICScraper{
		stateCode:      strings.ToUpper(stateCode),
		sessionFactory: sessionFactory,
	}
}

// StateCode returns the two-letter state code this scraper handles.
func (s *NAICScraper) StateCode() string { return s.stateCode }

// ManualLookupURL returns the URL for manual license lookups on the NAIC SBS portal.
func (s *NAICScraper) ManualLookupURL() string {
	return "https://sbs.naic.org/solar/external/pages/#/search/licensee/search"
}

// LookupByName searches for licenses by first and last name.
func (s *NAICScraper) LookupByName(ctx context.Context, firstName, lastName string) ([]LicenseResult, error) {
	params := url.Values{
		"jurisdiction": {s.stateCode},
		"searchType":   {"Licensee"},
		"entityType":   {"IND"},
		"firstName":    {firstName},
		"lastName":     {lastName},
	}
	return s.query(ctx, params)
}

// LookupByNPN searches for licenses by National Producer Number.
func (s *NAICScraper) LookupByNPN(ctx context.Context, npn string) ([]LicenseResult, error) {
	params := url.Values{
		"jurisdiction": {s.stateCode},
		"searchType":   {"Licensee"},
		"entityType":   {"IND"},
		"npn":          {npn},
	}
	return s.query(ctx, params)
}

// LookupByLicenseNumber searches for licenses by license number.
func (s *NAICScraper) LookupByLicenseNumber(ctx context.Context, licenseNumber string) ([]LicenseResult, error) {
	params := url.Values{
		"jurisdiction":  {s.stateCode},
		"searchType":    {"Licensee"},
		"entityType":    {"IND"},
		"licenseNumber": {licenseNumber},
	}
	return s.query(ctx, params)
}

// query performs the actual HTTP request to the NAIC API and parses the response.
func (s *NAICScraper) query(ctx context.Context, params url.Values) ([]LicenseResult, error) {
	session, err := s.sessionFactory()
	if err != nil {
		return nil, fmt.Errorf("naic: session error: %w", err)
	}

	reqURL := "https://services.naic.org/api/licenseLookup/search?" + params.Encode()
	log.Printf("NAIC query: %s", reqURL)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL, nil)
	if err != nil {
		return nil, fmt.Errorf("naic: request error: %w", err)
	}

	req.Header = http.Header{
		"Accept":            {"application/json"},
		"Origin":            {"https://sbs.naic.org"},
		"Referer":           {"https://sbs.naic.org/"},
		"Accept-Language":   {"en-US,en;q=0.9"},
		http.HeaderOrderKey: {"Accept", "Origin", "Referer", "Accept-Language"},
	}

	resp, err := session.Do(req)
	if err != nil {
		return nil, fmt.Errorf("naic: request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("naic: read body failed: %w", err)
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("naic: HTTP %d: %s", resp.StatusCode, string(body))
	}

	// Check for error response
	var errResp naicErrorResponse
	if json.Unmarshal(body, &errResp) == nil && errResp.UnexpectedError != "" {
		return nil, fmt.Errorf("naic: API error: %s", errResp.UnexpectedError)
	}

	// Parse as array of results
	var apiResults []naicAPIResponse
	if err := json.Unmarshal(body, &apiResults); err != nil {
		// Might be empty or unexpected format
		return []LicenseResult{{Found: false, State: s.stateCode}}, nil
	}

	if len(apiResults) == 0 {
		return []LicenseResult{{Found: false, State: s.stateCode}}, nil
	}

	var results []LicenseResult
	for _, r := range apiResults {
		// Parse active status from license type string
		active := strings.Contains(strings.ToLower(r.LicenseType), "active")

		// Parse residency
		resident := strings.EqualFold(r.Residency, "Yes")

		// Parse LOAs - replace <br/> with newline for readability
		loas := strings.ReplaceAll(r.LOAs, "<br/>", "\n")
		loas = strings.ReplaceAll(loas, "&lt;br/&gt;", "\n")

		// Parse name (format: "LAST, FIRST")
		fullName := r.Name

		results = append(results, LicenseResult{
			Found:          true,
			Active:         active,
			Resident:       resident,
			FullName:       fullName,
			LicenseNumber:  r.LicenseNumber.String(),
			NPN:            r.NPN,
			State:          s.stateCode,
			LicenseType:    r.LicenseType,
			Status:         r.LicenseType, // NAIC doesn't have separate status, it's embedded in type
			ExpirationDate: r.LicenseExpirationDate,
			IssueDate:      r.LicenseEffectiveDate,
			LOAs:           loas,
		})
	}

	return results, nil
}
